### DISCUSSION TITLE
Switch between documentation and diagnostic on a single line

### DISCUSSION THREAD

**Discussion by vfsoraki** [joaotavora/eglot#898] [Q&A] [answered]
I’ve recently switched from lsp-mode to Eglot. I think I’ve got things working pretty cool, but there’s one issue I can’t solve.

I can’t find how to see different diagnostics/documentations on a single like. Take this for example:

```
def some_func(%SomeStruct{} = a, b) do
```

If there’s a diagnostic message for this line, and I hove over `def`, I see the diagnostic message (in mini buffer), but it’s replaced by the documentation quickly. Hitting `C-h .` only shows the documentation. I want to see the diagnostic message (like some dialyzer issue for this function).

They’re probably provided by different packages (`eldoc` and `flymake` if I’m not mistaken). I can use `flymake-show-buffer-diagnostics` to view all issues of the current buffer, but it’s doesn’t show the complete message. If I can view these messages fully, I think my problem is solved. I tried running `flymake-show-diagnostic` on that line, but it doesn’t show anything and writes `Nothing at point` in mini buffer.

So I suppose the question is, how can I view the current (full) message of `flymake` for the current line?

I’m sure this is Eglot specific. I was using Doom before switching to my own config, and I think it was using `flycheck` and I didn’t have this problem there.

**Comment by joaotavora** [comment, marked as answer]
I think you can ask `eldoc` to show flymake diagnostics as a documentation item when within Eglot, and you can optionally also tell Eldoc to display all its items at once.  By default, Eglot only shows the signature and hover info items. And by default it will show only one, the most important of the two at once.

Try this 100% untested snippet in your `.emacs`.

```lisp
(add-hook 'eglot-managed-mode-hook (lambda ()
  (push #'flymake-eldoc-function eldoc-documentation-functions)
  (setq eldoc-documentation-strategy #'eldoc-documentation-compose) ;optional
))
```

Personally, to solve your problem I bind `M-p` and `M-n` to `flymake-goto-prev-error` and `flymake-goto-next-error`.  When they land me on an error, they tell me the diagnostic information at point.  And this trumps eldoc.

**Reply by joaotavora** [reply]
Lacking any feeback for 8 days, marking my own answer as the answer.  If it's not the answer, let me know.

**Reply by vfsoraki** [reply]
Please do. I didn’t have time to test this. I will update when I can do this.

**Reply by aclonegeek** [reply]
I modified that snippet slightly, and the following works for me:

```elisp
(add-hook 'eglot-managed-mode-hook
          (lambda ()
            ;; Show flymake diagnostics first.
            (setq eldoc-documentation-functions
                  (cons #'flymake-eldoc-function
                        (remove #'flymake-eldoc-function eldoc-documentation-functions)))
            ;; Show all eldoc feedback.
            (setq eldoc-documentation-strategy #'eldoc-documentation-compose)))
```

**Reply by vfsoraki** [reply]
Thanks @joaotavora. Your provided hook works (it’s missing a closing parenthesis, for future readers), but for some reason it shows errors twice (twice in mini buffer **and also**  twice in eldoc buffer).

Thanks @aclonegeek, this works beautifully. It also made my other problem go away! The problem was the documentation didn’t show up when it was nested inside another call (think `func_a(func_b())`, docs for `func_b` wouldn’t show up).

Great great, thanks!

**Reply by rileyrg** [reply]
good to find. Is there an easy tweak to enable this to popup the error at point?
